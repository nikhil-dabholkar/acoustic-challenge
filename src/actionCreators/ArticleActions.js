import { LOAD_ARTICLE_DETAILS_DATA, LOAD_ALL_ARTICLES_DATA } from '../constants/constants';
const axios = require('axios');

export const fetchSelectedArticleData = (articleId) => {
    return async (dispatch) => {
        const response = await axios.get(`https://content-eu-4.content-cms.com/api/859f2008-a40a-4b92-afd0-24bb44d10124/delivery/v1/content/${articleId}`);
        console.log(response);
        dispatch({
            type: LOAD_ARTICLE_DETAILS_DATA,
            payload: response.data
        })
    }
}

export const fetchAllArticlesData = () => {
    return async (dispatch) => {
        const response = await axios.get(`https://run.mocky.io/v3/8421cb68-a9d9-4575-ad5d-b4396c864231/`);
        console.log(response);
        dispatch({
            type: LOAD_ALL_ARTICLES_DATA,
            payload: response.data
        })
    }
}
