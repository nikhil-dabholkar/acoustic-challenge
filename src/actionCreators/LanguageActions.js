import { SELECT_LANGUAGE_PREFERENCE } from '../constants/constants'

/**
 * This action will be called when user changes the language
 */
export const switchLanguagePreference = (payload) => ({
  type: SELECT_LANGUAGE_PREFERENCE,
  payload: payload
})
