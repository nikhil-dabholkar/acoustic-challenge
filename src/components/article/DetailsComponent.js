import React, { useEffect } from 'react';
import { FormattedMessage } from 'react-intl';
import { LoadingComponent } from '../index';

function DetailsComponent({match, fetchArticle, selectedArticle, error, loading}) {

    useEffect(() => {
        fetchArticle(match.params.id);//'fa9519d5-0363-4b8d-8e1f-627d802c08a8'
    }, []);

    if (error && !loading) {
        throw error;
    } else if (!error && loading) {
        return (<LoadingComponent></LoadingComponent>);
    } else {
        return (<React.Fragment>
            <div className="App">
                Details... 
                <FormattedMessage id="app.header" defaultMessage="English" />
                <div className="App">
                    <h1>Article: {JSON.stringify(selectedArticle)}</h1>
                </div>
            </div>
        </React.Fragment>); 
    }
}

export default DetailsComponent;
