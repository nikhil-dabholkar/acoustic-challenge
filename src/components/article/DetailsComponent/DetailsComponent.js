import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, FormattedDate, FormattedMessage } from 'react-intl';
import { LoadingComponent } from '../../index';
import { Tabs, Tab, Tooltip } from 'carbon-components-react';
import styles from './DetailsComponent.module.scss';
import { Document32 } from '@carbon/icons-react';

function DetailsComponent({ match, fetchArticle, selectedArticle, error, loading, intl }) {

    useEffect(() => {
        /**
         * Calls this action on first time load (componentDidMount)
         */
        fetchArticle(match.params.id);//'fa9519d5-0363-4b8d-8e1f-627d802c08a8'
    }, []);

    if (error && !loading) {
        throw error;
    } else if (!error && !loading && selectedArticle && selectedArticle.id) {
        const tooltipDate = <Tooltip direction="bottom" tabIndex={0} >
            <p>{intl.formatMessage({id: "detailsComponent.dateTooltip"})}</p>
        </Tooltip>;
        const tooltipTextForBodySection = <Tooltip direction="bottom" tabIndex={0} >
            <p>{intl.formatMessage({id: "detailsComponent.bodySectionTooltip"})}</p>
        </Tooltip>;
        return (<React.Fragment>
            <div data-testid="detailsComponent" className="container">
                <div className="row">
                    <div className="col">
                        <div className="row">
                            <div className={styles.label}></div>
                            <div className={styles.label + ' ' + styles.heading}><Document32 className={styles.iconSpacing} />{selectedArticle.name}</div>
                        </div>
                    </div>
                </div>
                <div className="row">
                    <div className="col-md-3">
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.topicTitle" defaultMessage="Topic title"></FormattedMessage></div>
                            <div className={styles.value}>{selectedArticle.elements.heading.value}</div>
                        </div>
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.authorName" defaultMessage="Author name"></FormattedMessage></div>
                            <div className={styles.value}>{selectedArticle.elements.author.value}</div>
                        </div>
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.date" defaultMessage="Date"></FormattedMessage> * {tooltipDate}</div>
                            <div className={styles.value}>
                                <FormattedDate day="numeric" month="long" year="numeric" hour="numeric" minute="numeric" value={new Date(selectedArticle.elements.date.value)}></FormattedDate>
                            </div>
                        </div>
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.leadImage" defaultMessage="Lead image"></FormattedMessage></div>
                            <div className={styles.value}>
                                <a href={"https://content-eu.goacoustic.com/api/authoring" + (selectedArticle.elements.mainImage.value.leadImage.asset.resourceUri.substring((selectedArticle.elements.mainImage.value.leadImage.asset.resourceUri.indexOf('/v1/')), selectedArticle.elements.mainImage.value.leadImage.asset.resourceUri.length))} target="_blank">{selectedArticle.elements.mainImage.value.leadImage.asset.fileName}</a>

                            </div>
                        </div>
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.leadImageCaption" defaultMessage="Lead image caption"></FormattedMessage></div>
                            <div className={styles.value}>{selectedArticle.elements.mainImage.value.leadImageCaption.value}</div>
                        </div>
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.LeadImageCredit" defaultMessage="Lead image credit"></FormattedMessage></div>
                            <div className={styles.value}>{selectedArticle.elements.mainImage.value.leadImageCredit.value}</div>
                        </div>
                    </div>
                    <div className="col-md-9">
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.profiles" defaultMessage="Profiles"></FormattedMessage></div>
                            <div className={styles.value}>
                                <div style={{ width: '100%', height: '100%' }}>
                                    <Tabs>
                                        <Tab href="#" id="tab-1" label={intl.formatMessage({id: "detailsComponent.default"})} >
                                            <div className="some-content">
                                                <img src="https://images.unsplash.com/photo-1588702547919-26089e690ecc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=750&q=80" />
                                            </div>
                                        </Tab>
                                        <Tab href="#" id="tab-2" label={intl.formatMessage({id: "detailsComponent.lead"})} >
                                            <div className="some-content">
                                                <img src="https://images.unsplash.com/photo-1588702547919-26089e690ecc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=550&q=80" />
                                            </div>
                                        </Tab>
                                        <Tab href="#" id="tab-3" label={intl.formatMessage({id: "detailsComponent.card"})} >
                                            <div className="some-content">
                                                <img src={"https://images.unsplash.com/photo-1588702547919-26089e690ecc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=200&q=80"} />
                                            </div>
                                        </Tab>
                                    </Tabs>
                                </div>
                            </div>
                        </div>

                    </div>
                </div>
                <div className="row">
                    <div className="col">
                        <div className="row">
                            <div className={styles.label}><FormattedMessage id="detailsComponent.textForBodySection" defaultMessage="Text for body section"></FormattedMessage> {tooltipTextForBodySection}</div>
                            <div className={styles.value}>{
                                selectedArticle.elements.body.values.map((paragraphItem, index) => {
                                    const extracted = paragraphItem.replace('<p>', '').replace('</p>', '');
                                    return (<p key={index} className={styles.textForBodySectionParagraph}>{extracted}</p>)
                                })
                            }</div>
                        </div>
                    </div>
                </div>
            </div>
        </React.Fragment>);
    } else {
        return (<LoadingComponent></LoadingComponent>); 
    }
}

DetailsComponent.propTypes = {
    intl: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired,
    fetchArticle: PropTypes.func.isRequired,
    selectedArticle: PropTypes.object,
    error: PropTypes.object,
    loading: PropTypes.bool
}

export default injectIntl(React.memo(DetailsComponent));
