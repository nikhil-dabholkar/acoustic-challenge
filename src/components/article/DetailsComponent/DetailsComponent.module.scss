.label {
    display: inline-block;
    font-weight: 600;
    color: #343334;
    width: 100%;
    text-align: left;
    margin: 10px 0px;
    font-size: 19px;
}
.value {
    display: inline-block;
    font-weight: 500;
    color: #343334;
    width: 100%;
    text-align: left;
    margin: 10px 0 20px 10px;
    font-size: 15px;
}
.heading {
    margin-bottom: 20px;
    font-weight: 800;
    font-size: 23px;
}
.iconSpacing {
    padding-right: 10px;
    padding-bottom: 5px;
}
.textForBodySectionParagraph {
    padding: 0  0 10px 0;
}

.fullWidthHeight {
    width: '100%';
    height: '100%';
}