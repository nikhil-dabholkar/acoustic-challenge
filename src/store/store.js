import rootReducer from '../reducer/index';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';

export const store = createStore(rootReducer, applyMiddleware(thunk))