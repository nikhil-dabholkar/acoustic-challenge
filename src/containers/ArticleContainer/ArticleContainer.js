import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchSelectedArticleData, fetchAllArticlesData, resetArticlesError } from '../../actionCreators/ArticleActions';
import { Route, useLocation } from "react-router-dom";
import { useHistory } from 'react-router-dom'
import { lazy, Suspense } from 'react';
import { LoadingComponent } from '../../components/index';
import { ErrorBoundaryWrapper } from '../../wrappers/index';

const ListComponent = lazy(() => import("../../components/article/ListComponent/ListComponent"));
const DetailsComponent = lazy(() => import("../../components/article/DetailsComponent/DetailsComponent"));

function ArticleContainer({ resetError, fetchArticle, fetchAllArticles, article }) {
    const history = useHistory();
    const location = useLocation();
    const errorBoundaryRef = useRef(null);

    useEffect(() => {
        /**
         * When ever the url changes, clear the error 
         */
        return history.listen((location) => {
            resetError();
        })
    }, [history])

    return (<React.Fragment>
        <ErrorBoundaryWrapper key={location.pathname} ref={errorBoundaryRef}>
            <div data-testid="articleContainer">
                <Route path="/articles" render={(props) => {
                    return (<Suspense fallback={<LoadingComponent />}>
                        <ListComponent loading={article.loading} error={article.error} allArticles={article.allArticles} fetchAllArticles={fetchAllArticles} {...props} />
                    </Suspense>);
                }} exact />
                <Route path="/articles/:id" render={(props) => {
                    return (<Suspense fallback={<LoadingComponent />}>
                        <DetailsComponent loading={article.loading} error={article.error} selectedArticle={article.selectedArticle} fetchArticle={fetchArticle} {...props} />
                    </Suspense>)
                }} exact />
            </div>
        </ErrorBoundaryWrapper>
    </React.Fragment>
    );
}

const mapStateToProps = state => {
    return ({
        article: state.articleReducer
    })
};

const mapDispatchToProps = {
    fetchArticle: fetchSelectedArticleData,
    fetchAllArticles: fetchAllArticlesData,
    resetError: resetArticlesError
}

ArticleContainer.propTypes = {
    resetError: PropTypes.func.isRequired,
    fetchArticle: PropTypes.func.isRequired,
    fetchAllArticles: PropTypes.func.isRequired,
    article: PropTypes.object.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(React.memo(ArticleContainer));
