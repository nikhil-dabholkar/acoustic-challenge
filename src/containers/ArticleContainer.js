import React from 'react';
import { connect } from 'react-redux';
import { fetchSelectedArticleData, fetchAllArticlesData } from '../actionCreators/ArticleActions';
import { switchLanguagePreference } from '../actionCreators/LanguageActions';
import { BrowserRouter as Router, Route, Switch, Redirect, Link } from "react-router-dom";
import { lazy, Suspense } from 'react';
import { LoadingComponent, LanguageSwitcherComponent } from '../components/index';
import { ErrorBoundaryWrapper } from '../wrappers/index';
import { Breadcrumb, BreadcrumbItem } from 'carbon-components-react';
import styles from './styles/ArticleContainer.module.scss';
import HeaderComponent from '../components/shared/HeaderComponent';
const ListComponent = lazy(() => import("../components/article/ListComponent"));
const DetailsComponent = lazy(() => import("../components/article/DetailsComponent"));
const ErrorComponent = lazy(() => import("../components/shared/ErrorComponent"));

function ArticleContainer({ language, switchLanguage, fetchArticle, fetchAllArticles, article }) {
    return (<React.Fragment>
        <Router>
            <HeaderComponent language={language} switchLanguage={switchLanguage}></HeaderComponent>
            <ErrorBoundaryWrapper>
                <Switch>
                    <Route path="/articles" render={(props) => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <ListComponent loading={article.loading} error={article.error} allArticles={article.allArticles} fetchAllArticles={fetchAllArticles} {...props} />
                        </Suspense>);
                    }} exact />
                    <Route path="/articles/:id" render={(props) => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <DetailsComponent loading={article.loading} error={article.error} selectedArticle={article.selectedArticle} fetchArticle={fetchArticle} {...props} />
                        </Suspense>)
                    }} exact />
                    <Route path="/" render={() => <Redirect to="/articles" />} exact></Route>
                    <Route path="*" render={() => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <ErrorComponent />
                        </Suspense>)
                    }} exact />
                </Switch>
            </ErrorBoundaryWrapper>
        </Router>

    </React.Fragment>
    );
}

const mapStateToProps = state => {
    return ({
        article: state.articleReducer,
        language: state.languageReducer
    })
};

const mapDispatchToProps = {
    fetchArticle: fetchSelectedArticleData,
    fetchAllArticles: fetchAllArticlesData,
    switchLanguage: switchLanguagePreference
}

export default connect(mapStateToProps, mapDispatchToProps)(ArticleContainer);
