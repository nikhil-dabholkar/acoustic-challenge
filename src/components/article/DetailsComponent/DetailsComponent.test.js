import React from 'react'
import DetailsComponent from './DetailsComponent'
import ArticleContainer from '../../../containers/ArticleContainer/ArticleContainer'
import { IntlProvider } from 'react-intl'
import { store } from '../../../store/store'
import { Provider } from 'react-redux'
import { render, cleanup, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import enUS from '../../../i18n/en_US.json'
import { fetchSelectedArticleData } from '../../../actionCreators/ArticleActions'

afterEach(cleanup)

function renderWithRouter (
  ui,
  { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  return {
    ...render(<Router history={history}>{ui}</Router>),
    history
  }
}

function rendererWithRouter (
  ui,
  { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  return {
    ...renderer.create(<Router history={history}>{ui}</Router>),
    history
  }
}

describe('DetailsComponent', () => {
  test('Check if DetailsComponent Renders on present article', async () => {
    const route = '/articles/fa9519d5-0363-4b8d-8e1f-627d802c08a8'
    renderWithRouter(<Provider store={store}><IntlProvider locale="en" messages={enUS}><ArticleContainer></ArticleContainer></IntlProvider></Provider>, { route })
    const lazyElement = await screen.findByTestId('detailsComponent')
    expect(lazyElement).toBeInTheDocument()
  })

  test('Check if matches DetailsComponent Snapshot', async () => {
    const route = '/articles/fa9519d5-0363-4b8d-8e1f-627d802c08a8'
    const tree = rendererWithRouter(<Provider store={store}><IntlProvider locale="en" messages={enUS}><DetailsComponent fetchArticle={fetchSelectedArticleData} match={{ params: { id: 'fa9519d5-0363-4b8d-8e1f-627d802c08a8' } }}></DetailsComponent></IntlProvider></Provider>, { route }).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
