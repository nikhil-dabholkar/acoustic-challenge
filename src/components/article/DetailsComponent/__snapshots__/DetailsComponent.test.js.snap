// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsComponent Check if matches DetailsComponent Snapshot 1`] = `
<div
  data-testid="loadingComponent"
>
  <div
    className="bx--loading-overlay"
  >
    <div
      aria-atomic="true"
      aria-labelledby="loading-id-4"
      aria-live="assertive"
      className="bx--loading"
    >
      <label
        className="bx--visually-hidden"
        id="loading-id-4"
      >
        Active loading indicator
      </label>
      <svg
        className="bx--loading__svg"
        viewBox="-75 -75 150 150"
      >
        <title>
          Active loading indicator
        </title>
        <circle
          className="bx--loading__stroke"
          cx="0"
          cy="0"
          r="37.5"
        />
      </svg>
    </div>
  </div>
</div>
`;
